equinox/nn/_embedding.py

```diff
@@ -89,3 +96,313 @@ def __call__(
+        *,
+        key: Optional[PRNGKeyArray] = None,
+        **kwargs,
+    ):
+        """**Arguments:**
+        `RotaryPositionalEmbedding` requires:
+
+        - `embedding_size`: Size of the token embeddings. Must be non-negative.
+        - `key`: Not used; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+        """
+        if embedding_size < 0:
+            raise ValueError("embedding_size must not be negative.")
+        self.embedding_size = embedding_size
+
+    @staticmethod
+    def negate_half(x: Float[Array, "seq_len embedding_size"]):
+        d_2 = x.shape[-1] // 2
+        return jnp.concatenate([x[..., :d_2], -x[..., d_2:]], axis=-1)
+
+    @staticmethod
+    def precompute_freqs_cis(
+        embedding_size: int, end: int, theta: float = 10000.0
+    ) -> Complex[Array, "end embedding_size/2"]:
+        freqs = 1.0 / (
+            theta ** (jnp.arange(0, embedding_size, 2)[jnp.newaxis, :] / embedding_size)
+        )
+        t = jnp.arange(end)
+        freqs_outer = jnp.outer(t, freqs)
+        freqs_cis = jnp.cos(freqs_outer) + jnp.sin(freqs_outer) * 1j
+        return freqs_cis
+
+    @jax.named_scope("eqx.nn.RotaryPositionalEmbedding")
+    def __call__(
+        self,
+        x: Float[Array, "seq_len embedding_size"],
+        *,
+        key: Optional[PRNGKeyArray] = None,
+    ) -> Float[Array, "seq_len embedding_size"]:
+        """**Arguments:**
+
+        - `x`: A JAX array of shape `(seq_len, embedding_size)`.
+        - `key`: Ignored; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+
+        **Returns:**
+
+        A JAX array of shape `(seq_len, embedding_size)`, with the rotary positional
+        encoding applied to the input.
+        """
+
+        seq_len, embedding_size = x.shape
+        if embedding_size != self.embedding_size:
+            raise ValueError(
+                f"x.shape[-1] must match self.embedding_size, "
+                f"but {x.shape[-1]} != {self.embedding_size}"
+            )
+        if embedding_size % 2 != 0:
+            raise ValueError(
+                f"x.shape[-1] must be even, but {x.shape[-1]} is not even."
+            )
+
+        neg_half_x = self.negate_half(x)
+
+        with jax.ensure_compile_time_eval():
+            if (embedding_size, seq_len) in internal_rope_embedding_cache:
+                freqs_cis = internal_rope_embedding_cache[(embedding_size, seq_len)]
+            else:
+                freqs_cis = self.precompute_freqs_cis(embedding_size, seq_len)
+                internal_rope_embedding_cache[(embedding_size, seq_len)] = freqs_cis
+
+        assert freqs_cis is not None, "freqs_cis must not be None."
+        freqs_real = jnp.tile(freqs_cis.real, (1, 2))
+        freqs_imag = jnp.tile(freqs_cis.imag, (1, 2))
+
+        x_rope = (x * freqs_real) + (neg_half_x * freqs_imag)
+        return jax.lax.stop_gradient(x_rope)
```

Comment:
I don't think we want to stop the gradient flowing through our KVQ heads! I'm guessing this is a holdover from when the frequencies were stored on the layer itself.

Reply:
Yes, this was a leftover. Since the layer has no more `Array`, it has no trainable parameters. If it still had the `freqs_cis` as a field - and if we didn't use `jax.lax.stop_gradient` it - probably (I'm a bit shaky on `jax.lax.stop_gradient`) - would have changed the array.

Reply:
Interestingly, this makes quite a difference. With `stop_gradient` it's around 20% faster (on my MacBook, so it doesn't mean all that much - basically from 8 it/s to 10it/s). But the overall performance of the trained transformer is basically the same, a tiny bit better when not using `stop_gradient`. As it makes semantically more sense to _not_ stop the gradients, I'll keep it as is. But it's an interesting find.